crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -554,21 +492,14 @@ where
             })
             .collect::<HashMap<_, _>>();
 
-        let mut piece_caches_capacity_used = vec![0u32; caches.backends.len()];
+        let mut piece_caches_capacity_used = vec![0u32; caches.backends_count()];
         // Filter-out piece indices that are stored, but should not be as well as clean
         // `inserted_piece_indices` from already stored piece indices, leaving just those that are
         // still missing in cache
-        caches
-            .stored_pieces
-            .extract_if(|key, _offset| piece_indices_to_store.remove(key).is_none())
-            .for_each(|(_piece_index, offset)| {
-                // There is no need to adjust the `last_stored_offset` of the `backend` here,
-                // as the free_offset will be preferentially taken from the dangling free offsets
-                caches.dangling_free_offsets.push_front(offset);
-            });
+        caches.free_stored_piece_if(|key, _offset| piece_indices_to_store.remove(key).is_none());
```

Comment:
This should be replaced with a use-case specific method as well, `free_stored_piece_if` is way too literal translation of what we had before and doesn't make a lot of sense as a standalone API, hopefully I have given enough examples already to explain what it should look like.